### ISSUE TITLE
Demuxing dual-indexed, single-end fastq files with indexes in header fails

### ISSUE THREAD

**Issue by rikrdo89** [brwnj/fastq-multx#21]
I'm trying this tool for the first time to demultiplex a fastq file `Undetermined_L001_R1.fastq.gz` from an illumina run that was done using dual-indexes, and single-end reads. I have been getting a `Segmentation fault ` as error for a number of things when running this tool. 

So I decided to test this tool using the scripts and data in the test folder in this repository. Whenever I use the parameter `-H` to specify that the indexes are in the header of the reads, the tool always fails if I supply only one fastq file as input. however, when I supply two fastq files as input (entering the same file twice), the program runs without issues. My lines of code are below, with an excerpt of the data.

This fails:
```
$ fastq-multx -H -B indexes.txt test.fastq -o test_%.fastq
Using Barcode File: indexes.txt
Segmentation fault (core dumped)
```

but this runs (using the input file twice, and suppressing one of the ourputs, sort of a work around):
```
$ fastq-multx -H -B indexes.txt test.fastq test.fastq -o  test_%.fastq -o n/a
Using Barcode File: indexes.txt
End used: start
Id      Count   File(s)
S1    17800     test_S1.fastq
S2    32100     test_S2.fastq
S3    3900      test_S3.fastq
...
```

The files I am using look like this:
```
==> index.txt <==
S1     TTACCGAC-CGTATTCG
S2     TCGTCTGA-TCAAGGAC
S3     TTCCAGGT-AAGCACTG
S4     TACGGTCT-GCAATGGA
S5     AAGACCGT-CAATCGAC

==> test.fastq <==
@A00929:83:HL75TDRXX:1:2101:13919:1047 1:N:0:TTACCGAC+CGTATTCG
CATATTGATAGTTCGCACAGGTAG
+
FFFFFFFFFFFFFFFFFFFFFFFF
@A00929:83:HL75TDRXX:1:2101:14009:1047 1:N:0:TCGTCTGA+TCAAGGAC
GTGCGTATCTATCAAAAATGTATA
+
```

I installed fastq-multx using conda in Ubuntu 20.04

**Comment by hepcat72** [comment]
It's been a very long time since I've worked in C, but I looked at the code and have a mild hunch...  Try duplicating the seq defline on the qual defline and see if you still get a segfault.

**Comment by hepcat72** [comment]
Oh wait.  I assumed this was the same as the the other related issue about barcodes in headers... WRT dual index...
```
            if (bcinheader) {
                ignore=getline(&q, &ignore_st, fin[i]);
                ignore=getline(&q, &ignore_st, fin[i]);
                ignore=getline(&q, &ignore_st, fin[i]);
                /// no dual barcode detection allowed
                getbcfromheader(s, &ns);
                printf("bc is %s\n", s);
            } else {
```
The comment in the code seems to possibly suggest that dual barcodes in fastq headers may not be supported?

**Comment by rikrdo89** [comment]
The tool works well and demultiplexes without issue  if I just repeat the input twice (and the output as well), irrespective of dual barcodes. I think this is probably a bug in the code.

**Comment by hepcat72** [comment]
Yeah. I gathered that.  I was just trying to hazard a guess as to why it fails in the case where it segfaults.  If I had more time, I would try and figure it out and fix the bug.

**Comment by duao42** [comment]
> 如果我只是重复输入两次（以及输出），该工具就可以正常工作，并且不会出现问题，而不管双条形码。我认为这可能是代码中的一个错误。

Thank you very much, it helps me

**Comment by duao42** [comment]
> The tool works well and demultiplexes without issue if I just repeat the input twice (and the output as well), irrespective of dual barcodes. I think this is probably a bug in the code.

Thank you very much, it helps me

**Comment by misrak** [comment]
But, I feel that will give you duplicated results, right? For example, the same read will be printed twice for that sample file. So, instead of expecting 1 Mb per sample I will get 2 Mb per sample.

**Comment by hepcat72** [comment]
I think I read this issue wrong back in 2023.  (I had thought you meant that your index read was in the single end file [as opposed to the barcode in the header].)  I don't ever run `fastq-multx` manually - I have a script that does that, so I don't have the interface memorized, but in thinking about it today... the answer seems obvious (though maybe I'm not seeing something).  Your scenario needs 3 input files: 1. the single end read (the sequences you want to demultiplex), 2. the first index read file, and 3. the second index read file.  The tool will split all the files, but you need at least as many files as you have barcodes.  If your second index read is somehow in the single end read sequence (i.e. you have 1 file that has the read you want AND the 2 index reads), then to use this tool, you need to cut out the index reads and make separate input files.  You could shortcut and just cut out the second index, but both index reads have to be at the beginning of the read.

**Comment by rikrdo89** [comment]
I think maybe the  syntax usage for demultiplexing Dual-indexed Single End reads should be better documented. I don't think it is obvious, at least to me, to have two input fastq files when using dual indexes. In the manual, the parameters for using the barcodes from the illumina header and the barcodes file are `-B -H` , but there is no mention about needing an extra fastq file.

Below is the help I get when using the tool for reference.

```
$ fastq-multx
Usage: fastq-multx [-g|-l|-B] <barcodes.fil> <read1.fq> [mate.fq] -o r1.%.fq [-o r2.%.fq] ...
Version: 1.4.2

Output files must contain a '%' sign which is replaced with the barcode id in the barcodes file.
Output file can be n/a to discard the corresponding data (use this for the barcode read)

The barcodes file (-B) looks like this [where '-NNNNNNNN' for a dual index is optional]:

sample1 ATGGTCCA-TTGAGGAC
sample2 GCCTAAGT-AAGCGTCA
...

The column delimiter may be a space or a tab.  The best matches in the index file(s) must be unique to a single sample's barcode(s) or the read(s) will be considered unmatched.

Default is to guess the -bol or -eol based on clear stats.

If -g is used, then it's parameter is an index lane, and frequently occuring sequences are used.

If -l is used then all barcodes in the file are tried, and the *group* with the *most* matches is chosen.

Grouped barcodes file (-l or -L) looks like this [where '-NNNNNNNN' for a dual index is optional]:

sample1 ATGGTCCA-TTGAGGAC group1
sample1 GCCTAAGT-AAGCGTCA group1
sample2 CTTAGTTC-CCAAGTAC group2
...

Mated reads, if supplied, are kept in-sync

Options:

-o FIL1     Output files (one per input, required)
-g SEQFIL   Determine barcodes from the indexed read SEQFIL
-l BCFIL    Determine barcodes from any read, using BCFIL as a master list
-L BCFIL    Determine barcodes from <read1.fq>, using BCFIL as a master list
-B BCFIL    Use barcodes from BCFIL, no determination step, codes in <read1.fq>
-H          Use barcodes from illumina's header, instead of a read
-b          Force beginning of line (5') for barcode matching
-e          Force end of line (3') for barcode matching
-t NUM      Divide threshold for auto-determine by factor NUM (1), > 1 = more sensitive
-G NAME     Use group(s) matching NAME only
-x          Don't trim barcodes off before writing out destination
-n          Don't execute, just print likely barcode list
-v C        Verify that mated id's match up to character C (Use ' ' for illumina)
-m N        Allow N mismatches in union of all indexes, unless -M is supplied. (1)
-M M        Allow N,M mismatches in indexes 1,2 respectively (see -m N) (not used)
-d N        Require a minimum distance of N between the best and next best (2)
-q N        Require a minimum phred quality of N to accept a barcode base (0)
```

**Comment by hepcat72** [comment]
Totally agree. Incidentally, this isn't my tool, but I have contributed to it, because it's the fastest one I've found. I found the usage difficult to understand as well, which is part of the reason I created [my own wrapper in perl](https://github.com/hepcat72/htseq2multx/).  I build the fastq-multx command here:

https://github.com/hepcat72/htseq2multx/blob/cf62a180eedc0e80c964005ef08a9febeb3b4c51/htseq2multx.pl#L411-L435

My usage might be easier to follow, but it's still a bit cryptic, if I'm being honest.  Once I got it to work, we incorporated it into our pipelines, and I never look at it.

If you can, I would suggest submitting a PR with a usage update.
